### PR TITLE
Enable C++ 2020 compatibility

### LINKED ISSUE #54058: [MSVC][/std:c++latest] godot build failed due to error C2666: 'Vector2i::operator !=': 2 overloads have similar conversions
### Godot version

Latest source version from master branch(0d5e13c)

### System information

Windows Server 2016 + latest MSVC

### Issue description

Hello,

The MSVC team recently added godot as part of RWC testing to detect compiler regressions. And we found some errors when building godot with '/std:c++latest'. Can you help confirm if it‘s a issue? This issue can be reproduced on the internal MSVC.

One other request: How can we show the command line information in the build log? I didn't find a useful scons option in https://docs.oracle.com/cd/E88353_01/html/E37839/scons-1.html. I cannot easily reproduce the issue without the compile command, 

### Steps to reproduce

1. open VS2019/VS2022 x64 tools command
2. git clone --recursive https://github.com/godotengine/godot.git godot
3. cd godot
4. **set _CL_=/std:c++latest**
5. python.exe -m pip install scons
6. scons --version
7. scons tools=no target=release

### Minimal reproduction project

**Build.log**: [Godot_build.log](https://github.com/godotengine/godot/files/7386526/Godot_build.log)

**Error info:**
platform\windows\display_server_windows.cpp(770): error C2666: 'Vector2i::operator !=': 2 overloads have similar conversions
F:\gitP\godotengine\godot\core/math/vector2.h(345): note: could be 'bool Vector2i::operator !=(const Vector2i &) const'
C:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\um\propkeydef.h(36): note: or       'int operator !=(const PROPERTYKEY &,const PROPERTYKEY &)'
C:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\shared\guiddef.h(197): note: or       'bool operator !=(const GUID &,const GUID &)'
F:\gitP\godotengine\godot\core/string/string_name.h(180): note: or       'bool operator !=(const char *,const StringName &)'
F:\gitP\godotengine\godot\core/string/string_name.h(178): note: or       'bool operator !=(const String &,const StringName &)'
F:\gitP\godotengine\godot\core/string/ustring.h(464): note: or       'bool operator !=(const wchar_t *,const String &)'
F:\gitP\godotengine\godot\core/string/ustring.h(463): note: or       'bool operator !=(const char *,const String &)'
F:\gitP\godotengine\godot\core/math/vector2.h(344): note: or 'bool Vector2i::operator ==(const Vector2i &) const' [rewritten expression '!(x == y)']
F:\gitP\godotengine\godot\core/math/vector2.h(246): note: or 'bool Vector2::operator ==(const Vector2 &) const' [synthesized expression '!(y == x)']
F:\gitP\godotengine\godot\core/string/ustring.h(461): note: or 'bool operator ==(const char *,const String &)' [rewritten expression '!(x == y)']
F:\gitP\godotengine\godot\core/string/ustring.h(462): note: or 'bool operator ==(const wchar_t *,const String &)' [rewritten expression '!(x == y)']
F:\gitP\godotengine\godot\core/string/string_name.h(177): note: or 'bool operator ==(const String &,const StringName &)' [rewritten expression '!(x == y)']
F:\gitP\godotengine\godot\core/string/string_name.h(179): note: or 'bool operator ==(const char *,const StringName &)' [rewritten expression '!(x == y)']

### PR DESCRIPTION
**Pull request by bradc6**
This change enables building with C++ 2020 without using any c++20 concepts and remaining compatible with C++2017. It also attempts to make the least number of changes.

_Production edit: Closes https://github.com/godotengine/godot/issues/54058_

### CODE REVIEW

**Comment by YuriSizov** [comment]
Thanks for opening a PR. Most of these changes don't make sense from the coding perspective though. Those types that you are making explicit should be comparable already, why would it be incompatible with C++ 2020?

File: editor/editor_settings.cpp [hunk 1]
```diff
@@ -959,7 +959,7 @@ void EditorSettings::setup_network() {
 			continue;
 		}
 		// Select current IP (found)
-		if (ip == current) {
+		if (ip == IPAddress(current)) {
```
Comment (bruvzg) [inline editor/editor_settings.cpp:962]: Since it's cast and checked multiple times, it's probably better to do: `String ip_string = String(ip);` one in the beginning of the loop and use it for all checks and `hint +=`.

**Review by bruvzg — COMMENTED** [review]
Which part of C++20 require all these explicit casts? All `Variant` types should be compatible.

File: modules/mono/mono_gc_handle.h [hunk 2]
```diff
@@ -49,7 +49,9 @@ struct GCHandleIntPtr {
 	_FORCE_INLINE_ bool operator==(const GCHandleIntPtr &p_other) { return value == p_other.value; }
 	_FORCE_INLINE_ bool operator!=(const GCHandleIntPtr &p_other) { return value != p_other.value; }
 
+#if not defined(__cplusplus) or (__cplusplus < 202002L)
 	GCHandleIntPtr() = delete;
+#endif
```
Comment (neikeq) [inline modules/mono/mono_gc_handle.h:54]: Is this no longer supported in C++ 20?

File: core/variant/variant_parser.cpp [hunk 3]
```diff
@@ -1051,11 +1051,11 @@ Error VariantParser::parse_value(Token &token, Variant &value, Stream *p_stream,
 			bool got_bracket_token = false;
 			if (builtin_types.has(token.value)) {
 				array.set_typed(builtin_types.get(token.value), StringName(), Variant());
-			} else if (token.value == "Resource" || token.value == "SubResource" || token.value == "ExtResource") {
+			} else if (token.value == Variant("Resource") || token.value == Variant("SubResource") || token.value == Variant("ExtResource")) {
```
[the remark this answers is not in the document]
Comment (KaylinFinke) [reply core/variant/variant_parser.cpp:1054]: > Thanks for opening a PR. Most of these changes don't make sense from the coding perspective though. Those types that you are making explicit should be comparable already, why would it be incompatible with C++ 2020?

over.match.oper 12.4.2.3

TL;DR
Rewritten candidates include the set of comparison operators with their parameters reversed. Instead of simply implicitly converting one type to another, which probably isn't what you want, in C++20 there are two candidates for these operators in the overload set and neither is a better match than the other.

File: modules/mono/mono_gc_handle.h (hunk 2 above)
Comment (KaylinFinke) [reply modules/mono/mono_gc_handle.h:54]: dcl.init.aggr

`no user-declared or ... constructors ` (Until C++11, Since C++20)

Adding a user defined constructor would make this class non-trivial.
Adding an initializer would make this class non-trivial.
This seems like a minimal change.

File: editor/editor_inspector.cpp [hunk 4]
```diff
@@ -2175,7 +2175,7 @@ bool EditorInspectorArray::can_drop_data_fw(const Point2 &p_point, const Variant
 	}
 	Dictionary dict = p_data;
 	int drop_position = _drop_position();
-	if (!dict.has("type") || dict["type"] != "property_array_element" || String(dict["property_array_prefix"]) != array_element_prefix || drop_position < 0) {
+	if (!dict.has("type") || dict["type"] != Variant("property_array_element") || dict["property_array_prefix"] != Variant(array_element_prefix) || drop_position < 0) {
```
Comment (KaylinFinke) [inline editor/editor_inspector.cpp:2178]: `dict["property_array_prefix"] != Variant(array_element_prefix)`

should read:
`String(dict["property_array_prefix"]) != String(array_element_prefix)`
perhaps given we explicitly called operator String on LHS previously? Assuming that's valid.

File: editor/editor_inspector.cpp [hunk 5]
```diff
@@ -2216,7 +2216,7 @@ void EditorInspectorArray::_notification(int p_what) {
 
 		case NOTIFICATION_DRAG_BEGIN: {
 			Dictionary dict = get_viewport()->gui_get_drag_data();
-			if (dict.has("type") && dict["type"] == "property_array_element" && String(dict["property_array_prefix"]) == array_element_prefix) {
+			if (dict.has("type") && dict["type"] == Variant("property_array_element") && dict["property_array_prefix"] == Variant(array_element_prefix)) {
```
Comment (KaylinFinke) [inline editor/editor_inspector.cpp:2219]: same comment as above.

File: editor/editor_settings.cpp (hunk 1 above)
Comment (bradc6) [reply editor/editor_settings.cpp:962]: > Since it's cast and checked multiple times, it's probably better to do: `String ip_string = String(ip);` one in the beginning of the loop and use it for all checks and `hint +=`.

@bruvzg  Should we keep this PR about the mechanics of being C++20 compatible only and make a subsequent PR that changes the behavior as requested? Maybe make a link from that PR to here; or proceed with making the change here?

File: editor/editor_inspector.cpp (hunk 4 above)
Comment (bradc6) [reply editor/editor_inspector.cpp]: Applied freedback

File: editor/editor_inspector.cpp (hunk 5 above)
Comment (bradc6) [reply editor/editor_inspector.cpp]: Applied freedback

File: scene/gui/menu_bar.cpp [hunk 6]
```diff
@@ -376,7 +376,7 @@ void MenuBar::_notification(int p_what) {
 			}
 
 			Vector2 pos = DisplayServer::get_singleton()->mouse_get_position() - mouse_pos_adjusted - get_global_position();
-			if (pos == old_mouse_pos) {
+			if (pos == Vector2(old_mouse_pos)) {
```
Comment (akien-mga) [inline scene/gui/menu_bar.cpp:379]: Sounds like `old_mouse_pos` should have its type changed instead, this is the only place it's used.

Comment (bradc6) [reply scene/gui/menu_bar.cpp:379]: @akien-mga Should we make a behavioral change here by changing the type or leave this change as a one for one mechanically and make a second PR with the recommended behavioral change? Either is cool; I'm not sure what is preferred.

Comment (YuriSizov) [reply scene/gui/menu_bar.cpp:379]: For an upstream PR it's better to address the source of the issue from the get-go, than to add ad-hoc casts like that. In other words, if we know how to fix the problem, there is no reason to not fix it and do something else, and then make another PR to actually fix it.

Comment (YuriSizov) [reply scene/gui/menu_bar.cpp:379]: This includes comments made by bruvzg, btw. It should be preferable to use narrower types than Variant and/or to specify the operator.

Comment (KaylinFinke) [reply scene/gui/menu_bar.cpp]: I'm afraid we may be discussing distinct issues that touch the same lines of code. I'm in no way attempting to fix logic errors or improve performance issues in the context of this change. There are cases where T is being constructed or converted from U, currently implicitly, which cannot be implicitly done in future versions of the language in these contexts. Suggesting that making this behavior explicit somehow doesn't address this problem, but changing which types were constructed would, is to say that both the existing code is wrong by using the incorrect types -- perhaps true, and using the correct types would itself fix the language compatibility problem -- it would not.

Yes, there are numerous instances of questionable behavior caused by implicit conversions especially with regard to the Variant class. They are not limited to comparison operators but are found in assignment, function calls, and wherever the Variant is used in a context that expects a different type. Preventing this behavior is a significant undertaking which requires marking Variant's conversion functions and ctors explicit and then deciding on the behavior you want at each point it was previously being converted from/to and either calling the constructor for the right type or changing the behavior independently at every location. This should itself probably be two steps: one change to make such methods explicit without a behavior change and one to evaluate if the existing conversions are correct. Resolving overload ambiguity in comparison functions is the tip of the iceberg with regards to potential misuse of implicit conversions with the Variant class.

Experience has taught me not to couple behavioral changes with refactoring and I have intentionally not done so here as regressing against such a combined change can be difficult. If there's desire to couple behavior changes with a change that resolves overload ambiguity in future versions of the specification, we can do that, with understanding that it does not resolve the issue that is unintended implicit conversions, doesn't prevent people from doing such things in the future, and that it changes behavior of the existing code. I'd strongly recommend against it though.

Comment (YuriSizov) [reply scene/gui/menu_bar.cpp]: Strictly speaking, the engine does not benefit from this refactoring, but it would benefit from fixing issues in its code. There is little reason to do a pass that doesn't address the issues, but possibly even hides them better just to preserve some purity and atomic nature of changes. If you are not comfortable adjusting this PR to address the issues highlighted by your research, we can wait for somebody else to do it (but in this case, please open a bug report, so the problem is visible).

> Resolving overload ambiguity in comparison functions is the tip of the iceberg

Let's not get ahead of ourselves. You've made changes that clearly point at specific issues, and you've been advised how to address those issues so that the PR can improve the engine codebase for everyone. There is nothing more to it.

File: modules/mono/mono_gc_handle.h (hunk 2 above)
Comment (bradc6) [reply modules/mono/mono_gc_handle.h:54]: @neikeq I imagine the question is what hard constraints are you looking for on GCHandleIntPtr?
For example "It needs to be trivial/aggregate"
